### PR TITLE
Mission bulk displacement

### PR DESCRIPTION
**Pull request by rubenp02**
## Description
A variety of soft bugfixes and feature additions to enable a few simple mission-wide movement options.

### Make entryCoordinate explicit in VisualMissionItem
Previously, VisualMissionItem::coordinate was implicitly treated as the entry coordinate. This was unsuitable for LandingComplexItem, where the main coordinate is the landing (exit), as it is the master/pivot for user editing and transformations. Because of this, setCoordinate operated on the landing point, while the coordinate getter still returned the final approach (entry).

Introduced an explicit entryCoordinate alongside exitCoordinate, and redefined coordinate to alias either the entry or exit, depending on which is considered the main coordinate for a given VisualMissionItem subclass. For LandingComplexItem, coordinate and exitCoordinate now refer to the landing point, while the final approach is exposed via entryCoordinate.

In addition, Qt signal wiring for these properties has been simplified by using connections set up during object construction, reducing manual signal emissions.

### Add editableAlt getter to VisualMissionItem
Introduced a new editableAlt getter on VisualMissionItem to expose a single source of truth for the primary user-editable altitude across all VisualMissionItem subclasses. This value is intentionally independent of altitude reference-frame (relative, absolute, terrain), as it is intended to be used in conjunction with the likewise frame-agnostic VisualMissionItem::applyNewAltitude method to provide an external interface for generic altitude offsetting.

### Implement setCoordinate for complex scan items

This also fixes a bug where the "Edit position" option in the hamburger menu of Corridor Scan items did nothing.

### Add mission item bulk displacement utilities
Introduced new Q_INVOKABLE helpers on MissionController to transform entire missions, including:
- Repositioning the mission around a new home coordinate
- Offsetting the mission in ENU space
- Rotating the mission around home

Optional flags allow takeoff and landing items to be included or excluded. Repositioning includes them by default, while offset and rotation do not. All operations require a valid planned home position to succeed.

### Add hamburger menu with mission-wide transforms
Added a hamburger menu to the Mission Settings item that provides options to offset, rotate, and reposition the entire mission.

Also ensured DropPanel instances destroy themselves when closed to avoid leaking dynamically-created menus.

### Fix EditPositionDialog init reposition

Opening EditPositionDialog was triggering coordinateChanged during initialization, which then caused mission reposition or item coordinate updates even when nothing actually moved. Ensured initialization does not emit, while keeping explicit user actions behaving the same.

Key changes:
- Make EditPositionDialogController.coordinate REQUIRED
- Do not emit on the initial invalid-to-valid coordinate assignment
- Warn and ignore attempts to set an invalid coordinate
- Route Geo/UTM/MGRS updates through setCoordinate for behaviour centralization
- Set coordinate via QML binding in EditPositionDialog declarations, instead of via createObject property maps (no longer supported)

### Add mission item bulk displacement unit tests

The 2 tests cover a valid transform case (reposition + vertical offset) and an invalid-home no-op case.

## Type of Change
- [X] Bug fix (non-breaking change that fixes an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/Build changes
- [ ] Other

## Testing
- [X] Tested locally
- [X] Added/updated unit tests _Note: Just minor updates to keep them working with the new changes, nothing new is tested._ 
- [X] Tested with simulator (SITL)
- [ ] Tested with hardware

### Platforms Tested
- [ ] Linux
- [X] Windows
- [ ] macOS
- [ ] Android
- [ ] iOS

### Flight Stacks Tested
- [ ] PX4
- [ ] ArduPilot

## Screenshots
imagen
imagen
imagen

## Checklist
- [X] I have read the [Contribution Guidelines](CONTRIBUTING.md)
- [X] I have read the [Code of Conduct](CODE_OF_CONDUCT.md)
- [X] My code follows the project's coding standards
- [ ] I have added tests that prove my fix/feature works
- [X] New and existing unit tests pass locally
---
By submitting this pull request, I confirm that my contribution is made under the terms of the project's dual license (Apache 2.0 and GPL v3).

### CODE REVIEW

File: src/MissionManager/MissionController.cc [hunk 1]
```diff
@@ -2504,6 +2505,139 @@ void MissionController::setCurrentPlanViewSeqNum(int sequenceNumber, bool force)
...
+            continue;
+        }
+
+        double distanceMeters = oldHome.distanceTo(oldCoord);
+        double azimuthDegrees = oldHome.azimuthTo(oldCoord);
+
+        QGeoCoordinate newCoord = newHome.atDistanceAndAzimuth(distanceMeters, azimuthDegrees);
+        item->setCoordinate(newCoord);
+    }
+
```
[the remark this answers is not in the document]
Comment (rubenp02) [reply src/MissionManager/MissionController.cc:2720]: _settingsItem is index 0 which we're iterating through.

File: src/MissionManager/MissionController.cc [hunk 2]
```diff
@@ -2504,6 +2505,139 @@ void MissionController::setCurrentPlanViewSeqNum(int sequenceNumber, bool force)
...
+            if (!item->specifiesCoordinate() && !item->specifiesAltitudeOnly()) {
+                continue;
+            }
+
+            if (!qIsNaN(item->editableAlt())) {
+                item->applyNewAltitude(item->editableAlt() + upMeters);
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (rubenp02) [reply src/MissionManager/MissionController.cc:2769]: Same point as before, we're actually iterating through the home position as well, which is at index 0 in the _visualItems.

File: src/MissionManager/MissionController.h [hunk 3]
```diff
@@ -176,6 +177,48 @@ class MissionController : public PlanElementController
...
+    /// home coordinate. Requires a valid planned home position; otherwise the
+    /// mission is not modified and a warning is logged.
+    /// @param newHome New coordinate for the home item
+    /// @param repositionTakeoffItems If true, items identified as takeoff items
+    ///                               (isTakeoffItem) will be repositioned
+    /// @param repositionLandingItems If true, items identified as landing items
+    ///                               (isLandCommand) will be repositioned
+    Q_INVOKABLE void repositionMission(const QGeoCoordinate& newHome,
+                                       bool repositionTakeoffItems = true,
+                                       bool repositionLandingItems = true);
```
[the remark this answers is not in the document]
Comment (rubenp02) [reply src/MissionManager/MissionController.h:196]: I think there's enough context with the line:

`@param newHome New coordinate for the home item`

File: src/QmlControls/MissionItemEditor.qml [hunk 4]
```diff
@@ -259,6 +260,262 @@ Rectangle {
...
+        }
+    }
+
+    Component {
+        id: editMissionPositionDialogComponent
+
+        EditPositionDialog {
+            id: missionEditPositionDialog
+
+            onCoordinateChanged: _missionController.repositionMission(coordinate)
```
[the remark this answers is not in the document]
Comment (rubenp02) [reply src/QmlControls/MissionItemEditor.qml:539]: This one's actually true, hadn't noticed this. However, because it's called with the home position, its a no-op. Same thing with the EditPositionDialog used for the individual item position editing.

**Comment by DonLakeFlyer** [comment]
I'm out of town for two weeks so will take me a bit to get to this. Quick Look, says great new stuff. Just from looking at ui pics I would say two things:
* It doesn't look like you are using Facts to back the entry fields. I say that without looking at the diff since you have units in the label instead of in the text field itself like everywhere else. You should use Facts with metadata for these so they work like everything else, with validation, unit conversion and whatnot.
* In the Plan UI I am working towards using the next FactTextFieldSlider control for most numeric entry so you don't need to type in numbers on tablets. You can see usage of that new control in the settings item ui.

**Comment by rubenp02** [comment]
> I'm out of town for two weeks so will take me a bit to get to this. Quick Look, says great new stuff. Just from looking at ui pics I would say two things:
> 
>     * It doesn't look like you are using Facts to back the entry fields. I say that without looking at the diff since you have units in the label instead of in the text field itself like everywhere else. You should use Facts with metadata for these so they work like everything else, with validation, unit conversion and whatnot.
> 
>     * In the Plan UI I am working towards using the next FactTextFieldSlider control for most numeric entry so you don't need to type in numbers on tablets. You can see usage of that new control in the settings item ui.

Yes, I'm not using facts at all besides in the "Edit mission position..." popup, which uses the EditPositionDialog QML component. They should definitely be used. The QML side of this PR in general is a bit half-baked and I think that it really wants a new QML type similar to EditPositionDialog. When I get around to it, I can also try adding the sliders, but I'm not sure if that would work well here besides for the rotation stuff (what should the range be for the offset? Do the sliders work well for values that can also be negative?).

Also, any thoughts on the bug Copilot found about repositionMission being called upon constructing the EditPositionDialog? Since it also happens in the existing code when editing individual item positions.

File: src/QmlControls/MissionItemEditor.qml (hunk 4 above)
[the remark this answers is not in the document]
Comment (DonLakeFlyer) [reply src/QmlControls/MissionItemEditor.qml:539]: It ends up being a no-op but it's still crappy. Can you change coordinate in the controller to be required and then:
* Don't emit if transitioning from initial state of coordinate invalid
* qCWarning if there is an attempt to set to invalid

Is that doable?

**Comment by DonLakeFlyer** [comment]
@rubenp02 Why don't you just deal with the functionality for this. Then after this goes through I'll deal with usability, metadata, sliders and whatnot as needed.

File: src/QmlControls/MissionItemEditor.qml (hunk 4 above)
[the remark this answers is not in the document]
Comment (rubenp02) [reply src/QmlControls/MissionItemEditor.qml:539]: Of course, will do at some point next week.

**Comment by rubenp02** [comment]
> @rubenp02 Why don't you just deal with the functionality for this. Then after this goes through I'll deal with usability, metadata, sliders and whatnot as needed.

That works for me. Should I remove the commit for the QML side of things or is it okay to merge this as is (when I get around to doing your other suggestion) for the time being? To have a placeholder GUI of sorts.

**Comment by DonLakeFlyer** [comment]
> To have a placeholder GUI of sorts.

Yup. I'll work from there and update

**Comment by DonLakeFlyer** [comment]
@rubenp02 Any updates on this?

**Comment by rubenp02** [comment]
> @rubenp02 Any updates on this?

Forgot about it, sorry. Only thing missing is the proper handling of onCoordinateChanged, no? And now a rebase as well. If there's nothing else, I'll get around to it tomorrow.

File: src/QmlControls/MissionItemEditor.qml (hunk 4 above)
[the remark this answers is not in the document]
Comment (DonLakeFlyer) [reply src/QmlControls/MissionItemEditor.qml:539]: This is the only change needed

[the remark this answers is not in the document]
Comment (rubenp02) [reply src/QmlControls/MissionItemEditor.qml:539]: Done, worth noting that setting EditPositionDialog coordinate via createObject property maps is no longer supported as the property is marked required. The way it's done now is more idiomatic anyway, so I don't see it as a big deal.

**Comment by rubenp02** [comment]
Should be ready. Reminder that the QML side of things is meant to be improved, maybe an issue should be created to track that?

**Comment by rubenp02** [comment]
At least a few of Copilot's comments seem legit. I'll check them out and do the necessary changes when I can, but it might be a couple days.

**Comment by rubenp02** [comment]
Anything changed with Copilot? These seem much better than the initial ones

**Comment by DonLakeFlyer** [comment]
> Anything changed with Copilot? These seem much better than the initial ones

I've seen it be whack-a-mole. You fix a set of what it wants, run it again and then you get more. I'm hoping Claude will be better with respect to that.

**Comment by DonLakeFlyer** [comment]
@rubenp02 ?

**Comment by rubenp02** [comment]
> @rubenp02 ?

Sorry for disappearing but I've just been busy. I actually think I have everything fixed, I just need to check it and push it. I'll do it ASAP.

**Comment by rubenp02** [comment]
@DonLakeFlyer Should be done! Finally :)

**Comment by DonLakeFlyer** [comment]
This is done offline, with a PX4 Fixed Wing vehicle...

I start with this:
Screenshot 2026-03-04 at 10 10 24 AM

I then do a rotate 45 degrees including takeoff/land. I get this:
Screenshot 2026-03-04 at 10 10 49 AM

That doesn't seem quire right to me?

**Comment by DonLakeFlyer** [comment]
I don't get what Edit Mission Position does and how it differs from offset.

**Comment by rubenp02** [comment]
> I don't get what Edit Mission Position does and how it differs from offset.

Edit Mission Position is for absolute mission displacements and does not support changing the altitude. Offset Mission is for relative mission displacements and does support it.

Basically, if you want to bulk move all items in your mission slightly, maybe change all altitudes at once while keeping their original relative separation, you use Offset Mission. It's probably the more useful of the 2. Edit Mission Position is for when you want to place the same mission somewhere else in the Earth and have it be flown the exact same. You can then use Offset Mission to refine the result.

In short, they have related but different use cases. Both options get a lot of use in my company, as we often have to fly the exact same trajectories in another location for homologation stuff, and sometimes with 2 fixed-wing drones that must always be within safe distance between each other, for which we use a small 3-axis separation using Offset Mission (trajectories are basically the same, but traffic control is made a lot easier).



> This is done offline, with a PX4 Fixed Wing vehicle...
> 
> I start with this: (...)
> 
> I then do a rotate 45 degrees including takeoff/land. I get this: (...)
>
> That doesn't seem quire right to me?

It looks correct to me. You can partially check this by undoing the rotation and seeing how you get back to the original plan. The pivot is the Launch item (as that's the Home position) and you can see that stayed in the same spot as expected, and then the southbound leg until WP 3 now points roughly to the southwest, so a clockwise 45º rotation. Then, the landing position also got rotated by that much, but since it is a complex item, it doesn't get internally rotated, as this note says:

imagen

We would need a new transform API for complex items if we were to support internal rotation, which would take a pivot and an angle, rather than just the `setCoordinate` this uses now. Not sure if this is worth it in general. For our use case (of our company), it certainly isn't.

The whole entry vs exit vs main coordinate thing this PR includes is also related to this. We want to transform the coordinate for the main primitive the user cares about, not whichever happens to be the `entryCoordinate`.

**Comment by DonLakeFlyer** [comment]
> Edit Mission Position is for absolute mission displacements and does not support changing the altitude. Offset Mission is for relative mission displacements and does support it.

Ah, I get it now. I'll add some text in there when I update this to help a little for morons like me who cant understand it.

> It looks correct to me.

Ah, good. It was kinda making my head hurt to try to understand whether it was working or not.

**Comment by DonLakeFlyer** [comment]
Can you take a look at this review from Claude:
# Code Review: `Ventor-Innovations-feature/mission-bulk-displacement`

**7 commits, 21 files, +723 / -66 lines**

## Summary

This branch adds mission-wide transform operations (reposition, offset, rotate) to MissionController, implements `setCoordinate` for complex scan items, introduces an `entryCoordinate` / `editableAlt` interface on `VisualMissionItem`, and exposes the transforms via a hamburger menu on the mission settings (item 0) in the plan view.

---

## Architecture & Design

**Positive:**
- The `entryCoordinate` / `exitCoordinate` / `coordinate` separation is a sound refactor. Complex items like `LandingComplexItem` genuinely have distinct entry and exit points, and the old code was ambiguously aliasing `coordinate()` to one of them.
- The `editableAlt()` pure virtual is a clean way to get at the "user-facing altitude" without knowing the AMSL conversion details.
- The three transform functions (`repositionMission`, `offsetMission`, `rotateMission`) have consistent patterns with proper guard clauses and logging.

---

## Issues Found

### 1. `MissionSettingsItem::editableAlt()` returns raw fact, not AMSL — Semantic mismatch

`src/MissionManager/MissionSettingsItem.h:84`: `editableAlt()` returns `_plannedHomePositionAltitudeFact.rawValue()`, and then `amslEntryAlt()` delegates to `editableAlt()`. But `amslEntryAlt()` was previously returning `_plannedHomePositionCoordinate.altitude()`, which is the AMSL altitude (set via terrain data, vehicle, etc.). These are not the same value — the raw fact is the user-edited altitude, while `_plannedHomePositionCoordinate.altitude()` includes terrain offsets. This changes the semantics of `amslEntryAlt()` for the home position, which could cause incorrect terrain-relative flight path calculations.

**Recommendation:** Verify whether the home coordinate altitude and the fact raw value are always in sync. If not, `amslEntryAlt()` should continue to use `_plannedHomePositionCoordinate.altitude()` and `editableAlt()` can be a separate concept.

### 2. `LandingComplexItem::coordinate()` aliased to `exitCoordinate()` — Breaking semantic change

`src/MissionManager/LandingComplexItem.h:92`: `coordinate()` now returns `exitCoordinate()` (the landing point) instead of `_finalApproachCoordinate`. This is a visible behavioral change — any existing code or QML binding to `coordinate` on a landing item will now get the landing point instead of the approach point. The `setCoordinate()` implementation moves the landing coordinate, which is consistent with the new `coordinate() == exitCoordinate()` meaning, but:
- The chain of connects (`exitCoordinateChanged → coordinateChanged`) means `coordinateChanged` fires when the **landing** point changes, not the approach point. Consumers expecting `coordinate` to be the "primary working coordinate" (approach) will break.
- The `repositionMission` code in MissionController uses `item->coordinate()` to compute the old position for distance/azimuth. Since `coordinate()` now returns the landing point, the entire landing complex item is repositioned relative to the landing point, which may or may not be the intention.

**Recommendation:** Add a comment documenting why `coordinate()` was changed to point at the landing coordinate. Consider whether `repositionMission` needs special handling for `LandingComplexItem`.

### 3. Signal loops risk in `LandingComplexItem`

`src/MissionManager/LandingComplexItem.cc:44-49`: The new connects form a chain:
```
finalApproachCoordinateChanged → entryCoordinateChanged
landingCoordinateChanged → exitCoordinateChanged
exitCoordinateChanged → coordinateChanged
```
And `_recalcFromCoordinateChange` is connected to both `finalApproachCoordinateChanged` and `landingCoordinateChanged`. While `_ignoreRecalcSignals` guards against infinite loops in recalc, the `coordinateChanged` signal could propagate unexpectedly if anything is connected to it that in turn modifies the landing coordinate. This is fragile.

**Recommendation:** Audit all consumers of `coordinateChanged` on `LandingComplexItem` to ensure no feedback loops. Consider using a single-shot guard or deferred signal emission.

### 4. `EditPositionDialogController` suppresses initial `coordinateChanged` — Unintuitive behavior

`src/QmlControls/EditPositionDialogController.cc:49-55`: The `setCoordinate` setter now silently swallows the first valid coordinate assignment (the `wasInvalid` check). Combined with the `REQUIRED` property annotation, this means QML bindings that set `coordinate:` in the component instantiation will **not** trigger `onCoordinateChanged`. This is an intentional fix for the "init reposition" bug, but it makes the API surprising — callers can't know they won't get a signal.

**Recommendation:** Document this behavior clearly in the header. Consider an alternative like having a separate `initialize(QGeoCoordinate)` method vs the `setCoordinate` property setter.

### 5. `offsetMission` skips home altitude but `repositionMission` doesn't skip home position — Inconsistency

In `src/MissionManager/MissionController.cc` `repositionMission`, the settings item (home) **is** repositioned via `setCoordinate`, since `MissionSettingsItem` passes the `specifiesCoordinate()` check. But in `offsetMission`, the altitude loop explicitly skips `_settingsItem`. This asymmetry is by design (home altitude is managed separately), but the horizontal offset in `offsetMission` delegates to `repositionMission`, which **will** move the home — so the entire mission including home moves horizontally but only non-home items change altitude. This is correct but non-obvious.

**Recommendation:** Add a comment in `offsetMission` explaining that horizontal offset implicitly moves home via `repositionMission`, while altitude deliberately excludes home.

### 6. Complex item `setCoordinate` translates polygon but doesn't mark dirty

`TransectStyleComplexItem::setCoordinate`, `CorridorScanComplexItem::setCoordinate`, `StructureScanComplexItem::setCoordinate`: These all translate the polygon/polyline path but none call `setDirty(true)` directly. The caller (`repositionMission`) does call `setDirty(true)` at the end, and `_surveyAreaPolygon.setPath()` likely triggers dirty propagation internally, so it may be fine — but it leaves the `setCoordinate` API not self-contained.

**Recommendation:** Either ensure `setPath` propagates dirty correctly (and add a comment), or have `setCoordinate` set dirty itself for robustness.

### 7. Duplicated translate-polygon pattern

The polygon translation logic in `TransectStyleComplexItem::setCoordinate`, `CorridorScanComplexItem::setCoordinate`, and `StructureScanComplexItem::setCoordinate` is nearly identical (compute distance/azimuth, iterate vertices, `atDistanceAndAzimuth`).

**Recommendation:** Extract a shared helper like `QGCMapPolygon::translate(distance, azimuth)` or a free function to avoid the triple copy-paste.

### 8. QML: `Number.fromLocaleString` can return `NaN` silently

In the offset/rotate dialogs in `src/QmlControls/MissionItemEditor.qml`, the code handles `NaN` by defaulting to 0, which is fine. But the `DoubleValidator` already restricts input, so the `isNaN` checks are belt-and-suspenders — no issue, just noting it's defensive.

### 9. No `rotateMission` test

The test file adds `_testMissionTransforms` (covering `repositionMission` and `offsetMission`) and `_testMissionTransformsInvalidHome`, but there is no test for `rotateMission`. This is a meaningful gap since the rotation math (azimuth + offset degrees) could have edge cases near poles or the antimeridian.

**Recommendation:** Add a `rotateMission` test case that verifies items rotate correctly around home.

### 10. `_previousCoordinate` still uses `coordinate()` — intentional?

At `src/MissionManager/MissionController.cc:2397`, `_previousCoordinate = simpleItem->coordinate()` was not changed to `entryCoordinate()`. For simple items, `coordinate() == entryCoordinate()`, so this is fine, but it breaks the convention established by the rest of the diff.

---

## Minor / Style

- **Brace style inconsistency**: The new `editableAlt()` implementations use opening-brace-on-same-line (`double Foo::editableAlt() const {`), while the surrounding code uses next-line braces. Match the existing style per `CODING_STYLE.md`.
- **Log prefix duplication**: `qCWarning(MissionControllerLog) << "MissionControllerLog: ..."` — the logging category already identifies the source; the prefix in the string is redundant.
- The QML `enabled: _missionController.plannedHomePosition.isValid` bindings in the hamburger menu buttons are clean.
- Making the hamburger visible for sequence 0 (`visible: missionItem.isCurrentItem`) is a nice UX touch.

---

## Verdict

The overall design is solid and the approach to separating entry/exit coordinates is the right call. The main concerns are:
1. The `MissionSettingsItem::amslEntryAlt()` semantic change (Issue #1) — potential regression in flight path altitude calculations.
2. The `LandingComplexItem::coordinate()` change (Issue #2) — needs careful validation that all downstream consumers handle the new meaning.
3. Missing `rotateMission` test coverage (Issue #9).

**Comment by rubenp02** [comment]
From memory:
- I'll check on `1` but I think I remember making that change for cleanliness and checking that it didn't affect functionality. 
- `2` is intentional and correct.
- I'll check on `3` as well, that was actually me blindly following the Copilot review if it's currently wrong, as the original version didn't do this.
- `4` is very much as intended, as onCoordinateChanged is emitted by the GUI side on dialog creation. Documentation wouldn't hurt though.
- `5`, again, is very much intentional, and this one's actually documented in the docstrings IIRC.
- `6` is probably right.
- I disagree on `7` in this particular case, yes there is duplicate logic but it's minimal and trivial and in this particular instance I think its simpler to just have it duplicated rather than figuring out where to extract the common logic with only 2 usages.
- `8` is wrong, locale and validator are both needed, this isn't a belt and suspenders move as much as Qt stupidity.
- `9` is probably right, I guess I'll add one.
- `10` is wrong, we 100% want it to use the "main" coordinate.
- Style nitpick is right, missed that one. But I think what it says about the logging is wrong. Just using `qCWarning(MissionControllerLog)` doesn't prefix the message with `"MissionControllerLog: ..."`. I'll check just in case. The other stuff it says here are good things, or am I misunderstanding it?

I don't think any of the really bad ones it says at the end are right, but again, I'll check everything and update. And the GitHub issues it's listing there are completely random no?

**Comment by DonLakeFlyer** [comment]
> * But I think what it says about the logging is wrong.

It's correct. Log output by default includes category and source location.

**Comment by rubenp02** [comment]
> > * But I think what it says about the logging is wrong.
> 
> It's correct. Log output by default includes category and source location.

Then every log line in the file is wrong 😅 I'll fix the ones regarding thsi PR ASAP, as well as the other issues.

**Comment by rubenp02** [comment]
Looking at this more closely:

> ### 1. `MissionSettingsItem::editableAlt()` returns raw fact, not AMSL — Semantic mismatch
> 
> `src/MissionManager/MissionSettingsItem.h:84`: `editableAlt()` returns `_plannedHomePositionAltitudeFact.rawValue()`, and then `amslEntryAlt()` delegates to `editableAlt()`. But `amslEntryAlt()` was previously returning `_plannedHomePositionCoordinate.altitude()`, which is the AMSL altitude (set via terrain data, vehicle, etc.). These are not the same value — the raw fact is the user-edited altitude, while `_plannedHomePositionCoordinate.altitude()` includes terrain offsets. This changes the semantics of `amslEntryAlt()` for the home position, which could cause incorrect terrain-relative flight path calculations.
> 
> **Recommendation:** Verify whether the home coordinate altitude and the fact raw value are always in sync. If not, `amslEntryAlt()` should continue to use `_plannedHomePositionCoordinate.altitude()` and `editableAlt()` can be a separate concept.

Unsure as to wether it could possibly desync, but I think it's the right call in terms of coding style anyway, so I'm applying this suggestion.

> ### 2. `LandingComplexItem::coordinate()` aliased to `exitCoordinate()` — Breaking semantic change
> 
> `src/MissionManager/LandingComplexItem.h:92`: `coordinate()` now returns `exitCoordinate()` (the landing point) instead of `_finalApproachCoordinate`. This is a visible behavioral change — any existing code or QML binding to `coordinate` on a landing item will now get the landing point instead of the approach point. The `setCoordinate()` implementation moves the landing coordinate, which is consistent with the new `coordinate() == exitCoordinate()` meaning, but:
> 
>     * The chain of connects (`exitCoordinateChanged → coordinateChanged`) means `coordinateChanged` fires when the **landing** point changes, not the approach point. Consumers expecting `coordinate` to be the "primary working coordinate" (approach) will break.
> 
>     * The `repositionMission` code in MissionController uses `item->coordinate()` to compute the old position for distance/azimuth. Since `coordinate()` now returns the landing point, the entire landing complex item is repositioned relative to the landing point, which may or may not be the intention.
> 
> 
> **Recommendation:** Add a comment documenting why `coordinate()` was changed to point at the landing coordinate. Consider whether `repositionMission` needs special handling for `LandingComplexItem`.

This change was intentional, and there was already such a comment in `LandingComplexItem.cc`.

> ### 3. Signal loops risk in `LandingComplexItem`
> 
> `src/MissionManager/LandingComplexItem.cc:44-49`: The new connects form a chain:
> 
> ```
> finalApproachCoordinateChanged → entryCoordinateChanged
> landingCoordinateChanged → exitCoordinateChanged
> exitCoordinateChanged → coordinateChanged
> ```
> 
> And `_recalcFromCoordinateChange` is connected to both `finalApproachCoordinateChanged` and `landingCoordinateChanged`. While `_ignoreRecalcSignals` guards against infinite loops in recalc, the `coordinateChanged` signal could propagate unexpectedly if anything is connected to it that in turn modifies the landing coordinate. This is fragile.
> 
> **Recommendation:** Audit all consumers of `coordinateChanged` on `LandingComplexItem` to ensure no feedback loops. Consider using a single-shot guard or deferred signal emission.

No merit to this one, there are no feedback loops in the current code.

> ### 4. `EditPositionDialogController` suppresses initial `coordinateChanged` — Unintuitive behavior
> 
> `src/QmlControls/EditPositionDialogController.cc:49-55`: The `setCoordinate` setter now silently swallows the first valid coordinate assignment (the `wasInvalid` check). Combined with the `REQUIRED` property annotation, this means QML bindings that set `coordinate:` in the component instantiation will **not** trigger `onCoordinateChanged`. This is an intentional fix for the "init reposition" bug, but it makes the API surprising — callers can't know they won't get a signal.
> 
> **Recommendation:** Document this behavior clearly in the header. Consider an alternative like having a separate `initialize(QGeoCoordinate)` method vs the `setCoordinate` property setter.

This is documented in the implementation file. I'd argue it's an implementation detail. Wouldn't put it in the header.

> ### 5. `offsetMission` skips home altitude but `repositionMission` doesn't skip home position — Inconsistency
> 
> In `src/MissionManager/MissionController.cc` `repositionMission`, the settings item (home) **is** repositioned via `setCoordinate`, since `MissionSettingsItem` passes the `specifiesCoordinate()` check. But in `offsetMission`, the altitude loop explicitly skips `_settingsItem`. This asymmetry is by design (home altitude is managed separately), but the horizontal offset in `offsetMission` delegates to `repositionMission`, which **will** move the home — so the entire mission including home moves horizontally but only non-home items change altitude. This is correct but non-obvious.
> 
> **Recommendation:** Add a comment in `offsetMission` explaining that horizontal offset implicitly moves home via `repositionMission`, while altitude deliberately excludes home.

Again, already documented in what I consider to be the right place (the header in this case). In the implementation you already see what's going on in the code...

> ### 6. Complex item `setCoordinate` translates polygon but doesn't mark dirty
> 
> `TransectStyleComplexItem::setCoordinate`, `CorridorScanComplexItem::setCoordinate`, `StructureScanComplexItem::setCoordinate`: These all translate the polygon/polyline path but none call `setDirty(true)` directly. The caller (`repositionMission`) does call `setDirty(true)` at the end, and `_surveyAreaPolygon.setPath()` likely triggers dirty propagation internally, so it may be fine — but it leaves the `setCoordinate` API not self-contained.
> 
> **Recommendation:** Either ensure `setPath` propagates dirty correctly (and add a comment), or have `setCoordinate` set dirty itself for robustness.

This one lacks merit as well. `setDirty` is omitted where dirty propagation is handled by the connections set up in the constructor. I haven't found any path where geometry changes without the dirty state getting set.

> ### 7. Duplicated translate-polygon pattern
> 
> The polygon translation logic in `TransectStyleComplexItem::setCoordinate`, `CorridorScanComplexItem::setCoordinate`, and `StructureScanComplexItem::setCoordinate` is nearly identical (compute distance/azimuth, iterate vertices, `atDistanceAndAzimuth`).
> 
> **Recommendation:** Extract a shared helper like `QGCMapPolygon::translate(distance, azimuth)` or a free function to avoid the triple copy-paste.

As I said in my first impressions comment, I'd consider this overkill for the short code and low number of usages a helper would have.

> ### 8. QML: `Number.fromLocaleString` can return `NaN` silently
> 
> In the offset/rotate dialogs in `src/QmlControls/MissionItemEditor.qml`, the code handles `NaN` by defaulting to 0, which is fine. But the `DoubleValidator` already restricts input, so the `isNaN` checks are belt-and-suspenders — no issue, just noting it's defensive.

Same opinion as in my first impressions.

> ### 9. No `rotateMission` test
> 
> The test file adds `_testMissionTransforms` (covering `repositionMission` and `offsetMission`) and `_testMissionTransformsInvalidHome`, but there is no test for `rotateMission`. This is a meaningful gap since the rotation math (azimuth + offset degrees) could have edge cases near poles or the antimeridian.
> 
> **Recommendation:** Add a `rotateMission` test case that verifies items rotate correctly around home.

I agree, added.

> ### 10. `_previousCoordinate` still uses `coordinate()` — intentional?
> 
> At `src/MissionManager/MissionController.cc:2397`, `_previousCoordinate = simpleItem->coordinate()` was not changed to `entryCoordinate()`. For simple items, `coordinate() == entryCoordinate()`, so this is fine, but it breaks the convention established by the rest of the diff.

Opinion unchanged from my first impressions.

> ## Minor / Style
> 
>     * **Brace style inconsistency**: The new `editableAlt()` implementations use opening-brace-on-same-line (`double Foo::editableAlt() const {`), while the surrounding code uses next-line braces. Match the existing style per `CODING_STYLE.md`.

Fixed.

>     * **Log prefix duplication**: `qCWarning(MissionControllerLog) << "MissionControllerLog: ..."` — the logging category already identifies the source; the prefix in the string is redundant.

Fixed.

I've applied the changes I agree with. LMK if there's anything else you'd like from here, if perhaps you don't agree with some of my pushback.

**Comment by DonLakeFlyer** [comment]
Thanks, Ruben. Hoping to get this merged in today. Just want to take another quick look first.
